### ISSUE TITLE
HydroShare Schema.org Implementation

### ISSUE THREAD

**Issue by hyi** [hydroshare/hydroshare#3618]
Update JSON-LD implementation in HydroShare embedded in resource landing page to support mapping to schema.org metadata elements as complete as we can for better discovery of HydroShare resources by search engines, while ensuring compliance and compatibility to DataONE. For detailed design and background info, refer to https://gist.github.com/horsburgh/29c44a4b526603bc174605189d0001e3 created by @horsburgh.

**Comment by hyi** [comment]
@horsburgh you mentioned ```inLanguage``` element in schema.org should be mapped to HydroShare's language metadata element which implements language code as specified here: http://www.loc.gov/standards/iso639-2/php/code_list.php and defined in HS code here: https://github.com/hydroshare/hydroshare/blob/develop/hs_core/languages_iso.py. There is a little discrepancy between your example JSON for this element mapping, i.e., ```"inLanguage": "en-US"``` and the language metadata element code in HydroShare, i.e., ```inLanguage": "eng"```. So I want to confirm with you to make sure mapping this as ```inLanguage": "eng"``` (rather than ```"inLanguage": "en-US"```) should be sufficient to comply with schema.org.

**Comment by hyi** [comment]
@horsburgh I have done the first pass of Schema.org implementation and created a PR here https://github.com/hydroshare/hydroshare/pull/3705 for code review and discussion as needed. I also deployed the work to playground for you to test. However, since I added the code to check DEBUG setting and only show json-ld script on resource landing pages when DEBUG is false which only happens on the production server, you would not be able to see the json-ld schema.org markup unless I remove the conditional check for DEBUG flag on playground for your testing purpose. So I did that, i.e., removed the conditional check for DEBUG flag for your testing purpose for now. After you are done with testing, the DEBUG flag conditional check will ensure no schema.org json-ld markup implementation shows up on any development server, but only shows up on our production server. I did incorporate the info from the latest DataONE response in this first pass of implementation based on my understanding, and can make any further changes as you deem necessary once you get a chance to look deeper into Dave's response. It looks like we also need to incorporate the checksum of the resource bag zip in the schema.org markup, which I think would require some thinking through to implement mainly not to incur extra performance overhead for computing checksum of the resource bag as computing checksum of a large resource bag would be time-consuming. In any case, please test the current implementation out on playground and let me know any further changes to be made.

**Comment by horsburgh** [comment]
Awesome. Thanks @hyi - it will likely be next week before I get to this, but I'm glad to see the progress!

**Comment by horsburgh** [comment]
@hyi - I've done an initial round of testing and found several issues. The last one broke the resource I was testing with and so I'm going to post these here for your to fix and then I can continue testing when you've got these sorted out:

- [ ] 1. There's lots of unnecessary white space between elements in the code of the JSON. This could be cleaned up to be more compact

- [ ] 2. Comma after the "url" element of a "creator" should be replaced by a colon

``` "url", "https://www.hydroshare.org/user/6/" ```

should be:

``` "url": "https://www.hydroshare.org/user/6/" ```

- [ ] 3. The "url" element is being included for an author/"creator" that has no "url" and it is hard coded as:

"url", "https://www.hydroshare.org"

This is incorrect because it shouldn't be there and because it has the same issue with the comma instead of the colon between key and value.

- [ ] 4. There is a missing comma after "name" element of the "funder" that causes an error in validation

- [ ] 5. The "url" element for an organization that is an author is incorrectly listed as a blank string even though I've added a URL for that organization:

"url": ""

should be

"url": "http://uwrl.usu.edu"

- [ ] 6. The URL of "additionalType" element is not hardcoded to http://www.hydroshare.org/ but instead uses "playground.hydroshare.org". Not sure this is really an issue.

"additionalType": "http://playground.hydroshare.org/terms/CompositeResource",

to be consistent with the other URLs should be:

"additionalType": "http://www.hydroshare.org/terms/CompositeResource",

- [ ] 7. The "inLanguage" element uses "eng" instead of "en-US". Schema.org suggests the use of one of the language encodings from https://tools.ietf.org/html/bcp47. The correct language tag from this vocabulary is "en-US". I think it's probably OK to just set this because HydroShare only uses English, and doesn't have any way to change the "language" element in our resource metadata. This should probably be changed in HydroShare proper, but that's an issue for another time.

- [ ] 8. Author names are inconsistently listed. For example, for an author that is a HydroShare user:

"name": "Horsburgh, Jeff",

versus for an author that is not a HydroShare user:

"name": "Bryce Mihalevich",

If I go back into edit mode for the resource and enter Bryce's name as "Mihalevich, Bryce", then it shows up in the Schema.org metadata as "Mihalevich, Bryce". I thought we created a rule for names in HydroShare that if there was no comma in the name, then it was treated as "First Last"? I think in our Schema.org implementation, we would want the names to appear as "First Middle Last" without a comma. I think this problem has already been solved for the display of names on the resource landing page.

- [ ] 9. Additional URLs for an author are not being shown. For example, if I add a URL for a homepage to my user profile, the only URL that is shown is my HydroShare user URL and not the URL to my homepage. Adding a Google Scholar identifier to my HydroShare user profile broke the resource entirely. Now I get a Django error when I try to access it.

**Comment by hyi** [comment]
@horsburgh I have addressed all issues listed above and also updated playground so you can take another round of testing now. A few notes below for your reference:

- Issue 1 regarding a lot of white spaces, I have tried to tidy it up as much as I can without compromising too much on the code readability in our template html code. See if it looks all right now. As a result of this effort to remove white spaces, I ended up moving creator element creation to back end to create JSON formatted creator element rather than doing the logic in the template file at the front end since every new conditional statement would create a new empty line in the output script. Although the creator element is much more compact now without extra empty lines, the pretty print json created in backend always has the last closing array tag displayed as the first character on the line without accommodating the indentation. I tried many things but could not get this to work, so I'll leave it at that as a small cost to pay for removing extra white lines as well as increasing code readability.

- Issue 6 regarding "additionalType" containing playground in the url. I think we don't have to worry about it since when it is deployed to production, it will show the current www.hydroshare.org, and JSON-LD script will not be present on any non-production site (I manually commented out the enforcement check in code on playground for testing purpose, otherwise, JSON-LD script will not show). I did not hard code www.hydroshare.org for this "AdditionalType" element but instead retrieve it from metadata.type.url since it is readily available without the need to manually create the url for other scenarios. 

- Issue 7 regarding  "inLanguage," I have hardcoded it to "en-US" for now rather than retrieving it from metadata.language.code. Not sure whether we should create an issue for either changing hydroshare metadata proper for this or do a mapping between our current language code to inLanguage tag mapping, or we don't have to worry about it now. 

Let me know any further issues you find once you get a chance to test on playground again.

**Comment by horsburgh** [comment]
@hyi - thanks for the work. This is mostly looking good. I found a couple more minor issues. Also, we need to add a checksum for the zipped bag to the distribution element. I just sent Dave V. an email about that for some clarification. Once I get that, I'll modify the Gist in GitHub with an example.

- [ ] 11. In the "identifier" element, "datacite:usersIdentifierScheme" should be "datacite:usesIdentifierScheme" - sorry, that was a typo in my example
- [ ] 12. In the "creator" element, the "affilitation" element for each creator is expressed as a list, but should be a single element because the person should have a single affiliation. Remove the square brackets.
- [ ] 13. Is it convention in JSON to include an element with a blank string that does not have a value? For example, for creator Bryce Mihalevich, a url element is included even though there is no value for that element: ``` "url": "" ```. Or should it be ``` "url": null ``` Or, should it be excluded?
- [ ] 14. dateModified in the "distribution" element needs to be modified to ISO format.
- [ ] 15. If a custom license/rightsStatement is used and a URL is entered for that rights statement, it is not included in the schema.org representation.
- [ ] 16. If the user includes a place name in their spatial coverage, it is not included in the schema.org encoding.
- [ ] 17. We need to add the checksum for the zipped bag as an identifier to the "distribution" element. I'll add an example to the Gist in GitHub.

**Comment by hyi** [comment]
@horsburgh Thanks for the thorough test and review. I have further addressed all your further comments except for the last on regarding checksum and also updated playground for you to take another quick pass. For the last comment regarding adding checksum for the zipped bag, it will not be straightforward to implement due to our on-demand bag creation workflow for performance reasons meaning the zipped bag is not created or updated until a bag download request has been received. Is it OK not to include bag checksum in the schema.org JSON LD representation embedded in resource landing page but instead include it as part of downloaded bags? Currently a downloaded bag includes checksum for each included file in the bag, but not for the bag itself. Is it sufficient without creating a separate checksum for the bag zip file itself? The bag can be verified using checksums for each individual file, i.e., if each included file is valid, the zipped bag is valid. Let me know your thoughts and how we should proceed with this.

**Comment by horsburgh** [comment]
@hyi - awesome. With regard to the checksum issue, having the calculated checksum for a Bag is a requirement of DataONE. I know that the checksums for individual files are inside the zipped bag. However, what we are making available to DataONE for cataloging/download is the full zipped Bag and so they need a checksum of that object. They won't unzip the file. The alternative is that we include every file in the resource and its attributes (including checksum) within the Schema.org "distribution" element, and I don't want to do that unless we absolutely have to because some resources may have hundreds of files in them.

Could we trigger bag creation and checksum calculation on the bag as a last step of the resource publication process? Right now, we don't allow published files to change and so bag creation for published resources should be a one-time function. Also, I don't think resources are being formally published at a high rate and so I'm doubting that this would incur a huge performance penalty.

**NOTE:** Dave has asked me to look into allowing some resource metadata updates after publication to make it easier to add links to published papers. These would be limited edits that may or may not need to go in the resourcemetadata.xml document. If we do choose to edit the resourcemetadata.xml document with these changes after publication, it would affect the Bags for published resources, which would then have to be regenerated and checksum recalculated.

**Comment by hyi** [comment]
@horsburgh thanks for your comment and clarification. Providing bag checksum only for published resources definitely simplify things to a great deal. What you suggested should work but it will take me a bit of time to implement it. I also need to write a data migration command to create the bag and checksum as needed for existing resources that have already been published. If we need to account for updating resourcemetadata.xml, and thus bag and checksum, after a published resource is potentially updated for some metadata by an admin, that will complicate things a bit and require extra implementation, but can be done if needed. I have to double check, but I think if an admin made this metadata change to a published resource, metadata modified flag probably would be set to TRUE, which would trigger a resourcemetadata.xml update and bag recreation if this published resource were downloaded afterwards. If this were the case, I would have to update the bag checksum as well in order to make the distributed bag pass DataONE validation. I will have to double check this to make sure downloaded bag and checksum correspond to each other.

**Comment by hyi** [comment]
@horsburgh I have finished the first pass of adding checksum for published resources and updated playground for you to further test. I added checksum for published resources as an identifier in distribution element. An example is shown below for your reference:
```
{
            "@id":"ds-00#checksum",
            "@type":["PropertyValue", "datacite:ResourceIdentifier"],
            "datacite:usesIdentifierScheme": { "@id": "datacite:urn" },
            "propertyID": "http://id.loc.gov/vocabulary/preservation/cryptographicHashFunctions/md5",
            "value": "urn:hash:md5:dcc2c7d55a1167e6df887393f1462542"
}
```
Since we are still waiting for the response and clarification from DataONE related to this, we can iterate on this if any change is needed. Please take a look at this on playground and let me know any issues you may find.

**Comment by hyi** [comment]
@horsburgh regarding potentially updating metadata after a resource is published, I did not implement anything to automatically update the bag and checksum for the published resource when its metadata is updated after the resource publication action. However, metadata_dirty flag will be set for this published resource, and if the resource is downloaded by someone, the bag and checksum will be updated automatically due to the metadata_dirty flag being set to True. A management command by an admin can also be run on the server to update the bag and checksum for the affected published resource as needed. I think this should suffice for now, but let me know if you have any suggestions or comments.

**Comment by horsburgh** [comment]
@hyi - thanks. Will have a look. I need to think through the full workflow of what will happen if metadata are updated after a resource is published. Right now, I'm thinking we might want to use a management command to make sure that the resource doesn't stay in the dirty bag state for a long time after metadata are updated (e.g., if nobody downloads for a while).

**Comment by horsburgh** [comment]
@hyi - I think we should modify the checksum encoding to match the suggestion in this issue: 

https://github.com/ESIPFed/science-on-schema.org/issues/66

The following is an example (showing the full "distribution" element). I've also updated the Gist

https://gist.github.com/horsburgh/29c44a4b526603bc174605189d0001e3

```javascript
{
   "distribution":{
      "@type":"DataDownload",
      "contentSize":"8.2 MB",
      "encodingFormat":"application/zip",
      "contentUrl":"https://www.hydroshare.org/hsapi/resource/481cfb738f2b44e28305813a04aede71/",
      "description":"Zipped BagIt Bag containing the HydroShare Resource",
      "dateModified":"2019-12-09T21:21:50.284113+00:00",
      "identifier":[
         {
            "@type":[
               "PropertyValue",
               "datacite:ResourceIdentifier"
            ],
            "datacite:usesIdentifierScheme":{
               "@id":"datacite:local-resource-identifier-scheme"
            },
            "propertyId":"UUID",
            "value":"uuid:481cfb738f2b44e28305813a04aede71"
         },
         {
            "@type":"PropertyValue",
            "additionalType":[
               "http://www.wikidata.org/entity/Q185235",
               "http://id.loc.gov/vocabulary/preservation/cryptographicHashFunctions/md5"
            ],
            "identifier":"md5:dcc2c7d55a1167e6df887393f1462542",
            "propertyID":"MD5",
            "value":"dcc2c7d55a1167e6df887393f1462542"
         }
      ]
   }
}
```

**Comment by hyi** [comment]
@horsburgh Thanks. I have updated the code to use the suggested checksum encoding and also updated playground. Please take a quick test run on playground again when you have a chance and let me know if everything looks good now.

**Comment by hyi** [comment]
@horsburgh Now that python3 upgraded version has been merged into develop branch, I have merged my schema.org feature branch with the latest python3 develop branch and rebuilt hydroshare on playground on top of python 3 base image. I tested things out and fixed a couple of issues. Now everything appears to work well on python 3 base. I know you are planning to do a final round of quick testing, so thought it'd be good to get it ready for the big python 3 upgrade base before you are doing this final round of testing.
